### PR TITLE
Email & Password Validation 

### PR DESCRIPTION
**Pull request by edgaradamyan1**

### CODE REVIEW

File: Books/AppDelegate.swift [hunk 1]
```diff
@@ -31,6 +31,50 @@ class AppDelegate: UIResponder, UIApplicationDelegate {
...
+  // MARK: - Core Data stack
+
+  lazy var persistentContainer: NSPersistentContainer = {
+      /*
+       The persistent container for the application. This implementation
+       creates and returns a container, having loaded the store for the
+       application to it. This property is optional since there are legitimate
+       error conditions that could cause the creation of the store to fail.
+      */
+      let container = NSPersistentContainer(name: "coreDatatesz")
```
Comment (nellyAslanyan) [inline Books/AppDelegate.swift]: Change to the correct name

File: Books/Cell/CarouselCell.swift [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//  CarouselCell.swift
+//  Books
+//
+//  Created by Edgar Adamyan on 20.12.24.
+//
+
+import UIKit
+
+class CarouselCell: UICollectionViewCell {
+  @IBOutlet weak var onboardingImage: UIImageView!
```
Comment (nellyAslanyan) [inline Books/Cell/CarouselCell.swift:12]: add empty lines to the end and start of the objects.

File: Books/HomeVIewController/Cells/BottomCollectionViewCell.swift [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+import UIKit
+
+class BottomCollectionViewCell: UICollectionViewCell {
+  
+  @IBOutlet weak var authorImage: UIImageView!
+  @IBOutlet weak var authorName: UILabel!
+  @IBOutlet weak var typeOfPerson: UILabel!
+  
+
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/Cells/BottomCollectionViewCell.swift:16]: extra line here

File: Books/HomeVIewController/HomeViewController.swift [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+
+import UIKit
+import CoreData
+
+class HomeViewController: UIViewController {
+  
+  let searchButton = UIBarButtonItem(image: UIImage(named: "search"), style: .plain, target: nil, action: #selector(searchBook))
+  let notificationButton = UIBarButtonItem(image: UIImage(named: "notification"), style: .plain, target: nil, action: #selector(pushNotification))
+  let context = (UIApplication.shared.delegate as? AppDelegate)?.persistentContainer.viewContext
+
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift:17]: too many extra lines

File: Books/HomeVIewController/HomeViewController.swift [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+//
+//  Created by Edgar Adamyan on 23.12.24.
+//
+
+import UIKit
+import CoreData
+
+class HomeViewController: UIViewController {
+  
+  let searchButton = UIBarButtonItem(image: UIImage(named: "search"), style: .plain, target: nil, action: #selector(searchBook))
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: should be private?

File: Books/HomeVIewController/HomeViewController.swift [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+    bottomCollectionView.register(UINib(nibName: "BottomCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: "BottomCollectionViewCell")
+    topCollectionView.delegate = self
+    topCollectionView.dataSource = self
+    bottomCollectionView.dataSource = self
+    bottomCollectionView.delegate = self
+    
+  }
+  
+  
+  @objc func searchBook(){
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: Add space between () and {.

File: Books/HomeVIewController/HomeViewController.swift [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+  @IBOutlet weak var topCollectionView: UICollectionView!
+  @IBOutlet weak var bottomCollectionView: UICollectionView!
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    title = "Home"
+  
+    navigationItem.leftBarButtonItem = searchButton
+    navigationItem.rightBarButtonItem = notificationButton
+    navigationItem.leftBarButtonItem?.tintColor = .black
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: Lets have separate private functions for these purposes.

File: Books/HomeVIewController/HomeViewController.swift [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+  
+  
+  @objc func searchBook(){
+    
+  }
+  
+  @objc func pushNotification(){
+    
+  }
+  @IBAction func addBook(_ sender: UIButton) {
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift:52]: Add line

File: Books/OnboardingController/OnboardingController.swift [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+  //  ViewController.swift
+  //  Books
+  //
+  //  Created by Edgar Adamyan on 20.12.24.
+  //
+
+  import UIKit
+
+class OnboardingController: UIViewController {
+
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift:11]: Extra line

File: Books/OnboardingController/OnboardingController.swift [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+  import UIKit
+
+class OnboardingController: UIViewController {
+  
+ 
+  @IBOutlet weak var pageControl: UIPageControl!
+  @IBOutlet weak var collectionView: UICollectionView!
+  
+  let titleText: [String] = ["Now reading books will be easier", "Your Bookish Soulmate Awaits", "Start Your Adventure"]
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift]: should be private.

File: Books/OnboardingController/OnboardingController.swift [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+      
+    }
+  }
+  
+  func stopCaroucel(){
+    timer.invalidate()
+    
+    
+  }
+  @IBAction func SignInController(_ sender: UIButton) {
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift]: Lest have grouped functions e.g. @IBActions together, private functions together etc.

File: Books/SignInViewController/SignInViewController.swift [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+
+class SignInViewController: UIViewController {
+  var createdUsers: [User] = []
+
+  @IBOutlet weak var passwordTextField: UITextField!
+  @IBOutlet weak var emailTextField: UITextField!
+  @IBOutlet weak var errorText: UILabel!
+  
+  override func viewDidLoad() {
+        super.viewDidLoad()
```
Comment (nellyAslanyan) [inline Books/SignInViewController/SignInViewController.swift]: Indentation issue here

File: Books/HomeVIewController/HomeViewController.swift (hunk 8 above)
Comment (edgaradamyan1) [reply Books/HomeVIewController/HomeViewController.swift:50]: added

File: Books/HomeVIewController/Cells/TopCollectionVIewCell.swift [hunk 13]
```diff
@@ -0,0 +1,20 @@
...
+//  TopCollectionVIewCell.swift
+//  Books
+//
+//  Created by Edgar Adamyan on 23.12.24.
+//
+
+import UIKit
+
+class TopCollectionVIewCell: UICollectionViewCell {
+  @IBOutlet weak var titleofBook: UILabel!
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/Cells/TopCollectionVIewCell.swift:12]: line missing here

File: Books/String+Extensions/Validation.swift [hunk 14]
```diff
@@ -0,0 +1,34 @@
...
+//  Validation.swift
+//  Books
+//
+//  Created by Edgar Adamyan on 21.12.24.
+//
+
+import Foundation
+
+extension String{
+  func isValidPassword() -> Bool{
```
Comment (nellyAslanyan) [inline Books/String+Extensions/Validation.swift]: new lines missing

File: Books/SignUpViewController/SignUpViewController.swift [hunk 15]
```diff
@@ -0,0 +1,103 @@
...
+  
+  @IBAction func signInButton(_ sender: UIButton) {
+    let vc = SignInViewController(nibName: "SignInViewController", bundle: nil)
+    navigationController?.pushViewController(vc, animated: true)
+
+    
+  }
+  
+  
+
```
Comment (nellyAslanyan) [inline Books/SignUpViewController/SignUpViewController.swift]: extra lines

File: Books/SignUpViewController/SignUpViewController.swift [hunk 16]
```diff
@@ -0,0 +1,103 @@
...
+      self.activityIndicator.stopAnimating()
+      self.activityIndicator.isHidden = true
+    })
+    
+    let newUser = User(name: name, email: email, password: password)
+    saveUsersToUserDefaults(newUser)
+         
+  }
+  
+  func saveUsersToUserDefaults(_ user: User){
```
Comment (nellyAslanyan) [inline Books/SignUpViewController/SignUpViewController.swift]: should be private

File: Books/SignUpViewController/SignUpViewController.swift [hunk 17]
```diff
@@ -0,0 +1,103 @@
...
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    nameTextField.delegate = self
+    emailTextField.delegate = self
+    passwordTextField.delegate = self
+    registerButtonOutlet.isEnabled = false
+    activityIndicator.isHidden = true
+    
+
```
Comment (nellyAslanyan) [inline Books/SignUpViewController/SignUpViewController.swift]: extra lines

File: Books/SignInViewController/SignInViewController.swift [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+    }
+    let users = fetchUsersFromUserDefault()
+    if users.first(where: {$0.email == email && $0.password == password}) != nil{
+      print("Login successful")
+      errorText.text = ""
+      
+      let storyboard = UIStoryboard(name: "Main", bundle: nil)
+      let vc = storyboard.instantiateViewController(identifier: "TabBarController")
+      (UIApplication.shared.connectedScenes.first as? UIWindowScene)?.windows.first?.rootViewController = vc
+    }else{
```
Comment (nellyAslanyan) [inline Books/SignInViewController/SignInViewController.swift]: add missing spaces

File: Books/OnboardingController/OnboardingController.swift [hunk 19]
```diff
@@ -0,0 +1,99 @@
...
+      cell.titleText.text = titleText[indexPath.row]
+      cell.descriptionText.text = textDescription[indexPath.row]
+      cell.onboardingImage.image = UIImage(named: images[indexPath.row])
+    }
+    return cell
+  }
+  func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+    .init(width: collectionView.frame.width, height: collectionView.frame.height)
+  }
+  func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift:97]: new lines here

File: Books/OnboardingController/OnboardingController.swift [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+    startCarousel()
+    
+    
+  }
+  @IBAction func testButton(_ sender: UIButton) {
+    let storyBoard = UIStoryboard(name: "Main", bundle: nil)
+    let vc = storyBoard.instantiateViewController(withIdentifier: "TabBarController")
+    (UIApplication.shared.connectedScenes.first as? UIWindowScene)?.windows.first?.rootViewController = vc
+  }
+  @IBAction func SignInController(_ sender: UIButton) {
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift:40]: and here

File: Books/Model/User.swift [hunk 21]
```diff
@@ -0,0 +1,14 @@
+//
+//  User.swift
+//  Books
+//
+//  Created by Edgar Adamyan on 22.12.24.
+//
+
+import Foundation
+
+struct User: Codable{
```
Comment (nellyAslanyan) [inline Books/Model/User.swift]: add space

File: Books/HomeVIewController/HomeViewController.swift [hunk 22]
```diff
@@ -0,0 +1,106 @@
...
+    
+    return UICollectionViewCell()
+  }
+  
+  func collectionView(_ collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAt indexPath: IndexPath) -> CGSize {
+    if collectionView == topCollectionView{
+      let width = collectionView.frame.width / 3
+      let height = collectionView.frame.height
+      return CGSize(width: width, height: height)
+    }else if collectionView == bottomCollectionView{
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: add space

File: Books/TabBarController.swift [hunk 23]
```diff
@@ -0,0 +1,27 @@
...
+   
+      homeViewController.tabBarItem = UITabBarItem(title: "Home", image: UIImage(named: "home"), selectedImage: UIImage(named: "home"))
+      
+      setViewControllers([homeViewController], animated: true)
+      tabBar.tintColor = UIColor(named: "customViolet")
+      
+
+    }
+
+
```
Comment (nellyAslanyan) [inline Books/TabBarController.swift]: Identation and lines issues

File: Books/String+Extensions/Validation.swift [hunk 24]
```diff
@@ -0,0 +1,35 @@
...
+//  Created by Edgar Adamyan on 21.12.24.
+//
+
+import Foundation
+
+extension String{
+  
+  func isValidPassword() -> Bool{
+    let patern = "^(?=.*[A-Za-z])(?=.*\\d)[A-Za-z\\d]{7,}"
+    do{
```
Comment (nellyAslanyan) [inline Books/String+Extensions/Validation.swift]: add spaces :D

File: Books/SignUpViewController/SignUpViewController.swift [hunk 25]
```diff
@@ -0,0 +1,95 @@
...
+    var users = fetchUsersFromUserDefault()
+    users.append(user)
+    if let enodedData = try? JSONEncoder().encode(users){
+      UserDefaults.standard.set(enodedData, forKey: "users")
+    }
+    print("Saved users: \(users)")
+  }
+  
+  
+  func fetchUsersFromUserDefault() -> [User]{
```
Comment (nellyAslanyan) [inline Books/SignUpViewController/SignUpViewController.swift]: add spaces

File: Books/SignUpViewController/SignUpViewController.swift [hunk 26]
```diff
@@ -0,0 +1,95 @@
...
+      self.activityIndicator.stopAnimating()
+      self.activityIndicator.isHidden = true
+    })
+    
+    let newUser = User(name: name, email: email, password: password)
+    saveUsersToUserDefaults(newUser)
+         
+  }
+  
+  private func saveUsersToUserDefaults(_ user: User){
```
Comment (nellyAslanyan) [inline Books/SignUpViewController/SignUpViewController.swift]: add spaces

File: Books/SignInViewController/SignInViewController.swift [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+      
+      let storyboard = UIStoryboard(name: "Main", bundle: nil)
+      let vc = storyboard.instantiateViewController(identifier: "TabBarController")
+      (UIApplication.shared.connectedScenes.first as? UIWindowScene)?.windows.first?.rootViewController = vc
+    }else {
+      errorText.text = "False Login or Password"
+    }
+  }
+  
+  func fetchUsersFromUserDefault() -> [User]{
```
Comment (nellyAslanyan) [inline Books/SignInViewController/SignInViewController.swift]: add spaces

File: Books/OnboardingController/OnboardingController.swift [hunk 28]
```diff
@@ -0,0 +1,102 @@
...
+    navigationController?.pushViewController(vc, animated: true)
+  }
+  
+  @IBAction func pageControlChanged(_ sender: UIPageControl) {
+    currentIndex = sender.currentPage
+    let indexPath = IndexPath(row: currentIndex, section: 0)
+    collectionView.scrollToItem(at: indexPath, at: .centeredHorizontally, animated: true)
+  }
+  
+  func startCarousel(){
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift]: add spaces

File: Books/OnboardingController/OnboardingController.swift [hunk 29]
```diff
@@ -0,0 +1,102 @@
...
+      
+      let indexPath = IndexPath(row: self.currentIndex, section: 0)
+      self.collectionView.scrollToItem(at: indexPath, at: .centeredHorizontally, animated: true)
+      
+      self.pageControl.currentPage = self.currentIndex
+      
+    }
+  }
+  
+  func stopCaroucel(){
```
Comment (nellyAslanyan) [inline Books/OnboardingController/OnboardingController.swift]: add spaces

File: Books/HomeVIewController/HomeViewController.swift [hunk 30]
```diff
@@ -0,0 +1,106 @@
...
+  }
+  
+  func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
+    if collectionView == topCollectionView{
+      let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "TopCollectionVIewCell", for: indexPath)
+      if let cell = cell as? TopCollectionVIewCell{
+        cell.priceOfBook.text = "$5"
+        return cell
+      }
+    }else if collectionView == bottomCollectionView{
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: add spaces

File: Books/HomeVIewController/HomeViewController.swift [hunk 31]
```diff
@@ -0,0 +1,106 @@
...
+  @IBOutlet weak var bottomCollectionView: UICollectionView!
+  
+  override func viewDidLoad() {
+    super.viewDidLoad()
+    title = "Home"
+    setNavigationItem()
+    configureTableVIew()
+  }
+  
+  func configureTableVIew(){
```
Comment (nellyAslanyan) [inline Books/HomeVIewController/HomeViewController.swift]: add spaces

File: Books/HomeVIewController/HomeViewController.swift (hunk 5 above)
Comment (nellyAslanyan) [reply Books/HomeVIewController/HomeViewController.swift]: Please recheck this one

**Review by nellyAslanyan — APPROVED** [review]

**Pull request merged**